### PR TITLE
Add context awareness to login cmd

### PR DESCRIPTION
**Pull request by axbarsan**
Closes https://github.com/giantswarm/giantswarm/issues/11793

This PR might seem huge, but it is a bit bloated by functions extracted to separate files and test files.

Added all the requirements from the issue, and on top of that, added support for auth token renewal.

## Preview

<details><summary>Using help flag</summary>

image

</details>

<details><summary>Logging in using K8s API URL</summary>

image

</details>

<details><summary>Logging in using Happa URL</summary>

image

</details>

<details><summary>Logging in using an invalid URL</summary>

image

</details>

<details><summary>Logging in using 'gs-ginger' context name</summary>

image

</details>

<details><summary>Logging in using 'gs-unknown' context name</summary>

image

</details>

<details><summary>Logging in using 'ginger' installation name</summary>

image

</details>

<details><summary>Logging in using 'unknown' installation name</summary>

image

</details>

<details><summary>Logging in using no arguments (already logged into `ginger`)</summary>

image

</details>

<details><summary>Logging in using no arguments (already logged in with a non-`gs-*` context)</summary>

image

</details>

### CODE REVIEW

File: cmd/login/auth.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
+package login
```
Comment (axbarsan) [inline cmd/login/auth.go:1]: This file's contents were extracted from the big `runner.go` file

File: cmd/login/command.go [hunk 2]
```diff
@@ -9,11 +9,36 @@ import (
...
 	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/giantswarm/kubectl-gs/pkg/middleware"
+	"github.com/giantswarm/kubectl-gs/pkg/middleware/renewtoken"
 )
 
 const (
-	name        = "login"
-	description = "Logs in into the Kubernetes cluster."
+	name             = "login [K8s API URL | Web UI URL | Existing GS Context Name]"
```
Comment (axbarsan) [inline cmd/login/command.go:18]: Command documentation

File: pkg/installation/info.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
+package installation
```
Comment (axbarsan) [inline pkg/installation/info.go:1]: Extracted from `installation.go`

File: pkg/middleware/middleware.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package middleware
```
Comment (axbarsan) [inline pkg/middleware/middleware.go:1]: Added the concept of middleware. These are meant to be used before running a command (cobra command `PreRunE` field) or after running it (cobra command `AfterRunE` field).

These can be used for checking if a user is logged in before executing any other command.

Right now they are used for refreshing the authentication token on the already selected Giant Swarm kubectl context.

File: pkg/middleware/renewtoken/renewtoken.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/giantswarm/kubectl-gs/pkg/kubeconfig"
+	"github.com/giantswarm/kubectl-gs/pkg/middleware"
+	"github.com/giantswarm/kubectl-gs/pkg/oidc"
+)
+
+// Middleware will attempt to renew the current context's auth info token.
+// If the renewal fails, this middleware will not fail.
+func Middleware(k8sConfigAccess clientcmd.ConfigAccess) middleware.Middleware {
```
Comment (axbarsan) [inline pkg/middleware/renewtoken/renewtoken.go:14]: Middleware for renewing the auth token

File: pkg/oidc/token.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package oidc
+
+import (
+	"github.com/giantswarm/microerror"
+	"golang.org/x/oauth2"
+)
+
+func ConvertTokenToRawIDToken(token *oauth2.Token) (string, error) {
```
Comment (axbarsan) [inline pkg/oidc/token.go:8]: Extracted from `oidc.go`

**Comment by marians** [comment]
Starting at the end, using the login command without arguments, with a current-context not created by kubectl-gs.

```
kgs login
Error: selected context non compatible error: The currently selected context is not compatible, please try to login using the Kubernetes API URL or the web UI URL.
Run 'kgs login --help' to learn more.
panic: "{\"kind\":\"selectedContextNonCompatibleError\",\"annotation\":\"The currently selected context is not compatible, please try to login using the Kubernetes API URL or the web UI URL.\\nRun 'kgs login --help' to learn more.\",\"stack\":[{\"file\":\"/Users/marian/git/giantswarm/kubectl-gs/cmd/login/runner.go\",\"line\":93},{\"file\":\"/Users/marian/git/giantswarm/kubectl-gs/cmd/login/runner.go\",\"line\":52},{\"file\":\"/Users/marian/git/giantswarm/kubectl-gs/cmd/login/runner.go\",\"line\":40},{\"file\":\"/Users/marian/git/giantswarm/kubectl-gs/main.go\",\"line\":57}]}"


goroutine 1 [running]:
main.main()
	/Users/marian/git/giantswarm/kubectl-gs/main.go:19 +0x101
```

My current-context was one that has not been created by kubectl-gs.

Can we avoid the panic output? That looks quite intimidating and doesn't add value to the ordinary user.

Also I'm thinking about how to improve the message. Suggestion:

`The current context '<name-of-context>' does not seem to belong to a Giant Swarm control plane. Please run 'kgs login --help' to find out how to log in to a particular control plane.`

**Comment by marians** [comment]
Just saw that the panic output is a global thing. Can we avoid this in general?

**Comment by marians** [comment]
Regarding the kubeconfig modifications. This is what's created for gauss:

```
- context:
    cluster: gauss
    user: marian-gauss
  name: gs-gauss
```

I think it would be good to also prefix the user and cluster name, to make it more obvious that the entries were created by kubectl-gs and to avoid overwriting something the user might have added otherwise.

Cluster name: `gs-gauss`

User name: `gs-gauss-marian`

**Comment by marians** [comment]
Trying to log in with using an API URL or host name as an argument fails

```
kgs login api.g8s.gauss.eu-central-1.aws.gigantic.io
kgs login https://api.g8s.gauss.eu-central-1.aws.gigantic.io
kgs login https://api.g8s.gauss.eu-central-1.aws.gigantic.io/
```

Error is always:

```
Error: cannot get installation info: make sure you're connected to the internet
panic: "{\"kind\":\"cannotGetInstallationInfo\",\"annotation\":\"make sure you're connected to the internet\",\"stack\":[{\"file\":\"/Users/marian/git/giantswarm/kubectl-gs/pkg/installation/info.go\",\"line\":23},{\"file\":\"/Users/marian/git/giantswarm/kubectl-gs/pkg/installation/installation.go\",\"line\":30},{\"file\":\"/Users/marian/git/giantswarm/kubectl-gs/cmd/login/runner.go\",\"line\":134},{\"file\":\"/Users/marian/git/giantswarm/kubectl-gs/cmd/login/runner.go\",\"line\":78},{\"file\":\"/Users/marian/git/giantswarm/kubectl-gs/cmd/login/runner.go\",\"line\":40},{\"file\":\"/Users/marian/git/giantswarm/kubectl-gs/main.go\",\"line\":57}]}"
```

File: cmd/login/command.go [hunk 7]
```diff
@@ -9,11 +9,36 @@ import (
...
+	shortDescription = "Logs into an installation's Kubernetes API"
+	longDescription  = `Log into an installation's Kubernetes API.
+
+You can use as an argument:
+  * Your installation's Kubernetes API URL as an argument (e. g. 'https://g8s.test.eu-west-1.aws.gigantic.io')
```
Comment (marians) [inline cmd/login/command.go:23]:
```suggestion
  * Your installation's Kubernetes API URL, e. g. 'https://g8s.test.eu-west-1.aws.gigantic.io'
```

Also I think our API host names usually start with `api.g8s.`.

File: cmd/login/command.go [hunk 8]
```diff
@@ -9,11 +9,36 @@ import (
...
+  * Your Web UI URL (e. g. 'https://happa.g8s.test.eu-west-1.aws.gigantic.io')
+  * An existing Giant Swarm specific 'kubectl' context name (e. g. 'gs-test')
```
Comment (marians) [inline cmd/login/command.go:25]:
```suggestion
  * An existing Giant Swarm specific kubectl context name (e. g. 'gs-test')
```

File: cmd/login/command.go [hunk 9]
```diff
@@ -9,11 +9,36 @@ import (
...
+`
+	examples = `  # See on which installation you're logged in currently.
+  kubectl gs login
+
+  # Log in using your K8s API URL.
+  kubectl gs login https://g8s.test.eu-west-1.aws.gigantic.io
```
Comment (marians) [inline cmd/login/command.go:31]: Again, please check whether this should be `api.g8s.`

File: cmd/login/runner.go [hunk 10]
```diff
@@ -65,170 +44,119 @@ func (r *runner) Run(cmd *cobra.Command, args []string) error {
...
+
+	return nil
+}
 
-		w.Header().Set("Content-Type", "text/html")
-		http.ServeContent(w, r, "", time.Time{}, successTemplate)
+func (r *runner) tryToReuseExistingContext() error {
+	if currentContext, isLoggedInWithKubeContext := isLoggedWithGSContext(r.k8sConfigAccess); isLoggedInWithKubeContext {
+		codeName := getCodeNameFromKubeContext(currentContext)
+		fmt.Fprint(r.stdout, color.GreenString("You are logged on installation '%s'.\n", codeName))
```
Comment (marians) [inline cmd/login/runner.go:88]:
```suggestion
		fmt.Fprint(r.stdout, color.GreenString("You are logged in to the control plane of installation '%s'.\n", codeName))
```

**Comment by marians** [comment]
image

The `Switched to context ...` suggests that the current-context before calling the command was a different one, which was not the case here. Let's avoid this, as it makes the user wonder about kubectl-gs' sanity or even their own.

In this case, ideally we would print `Context '...' is already selected.`, plus the final success line.

**Comment by marians** [comment]
We have some inconsistency to iron out between help text and other output.

E. g. when I call `kubectl gs login gs-gauss`, the output uses `kgs` in info like `Note: No need to pass the 'gs-' prefix. 'kgs login gauss' works fine.`.

But when I call `kgs login --help`, `kubectl gs` is used instead of `kgs`.

The straightforward fix is to replace `kubectl gs` with `kgs` in the help text. However, we should also prepend the examples with this line:

```
Note: 'kgs' is an alias for 'kubectl gs'.
```

**Comment by marians** [comment]
@axbarsan I'm done for now

**Comment by axbarsan** [comment]
@marian I made all the requested changes 👍 

As for getting rid of the huge error output, there's https://github.com/giantswarm/giantswarm/issues/11983 for tracking it.

About K8s API URL:
* `api.g8s.ginger.*` - This is the Giant Swarm API URL format.
* `g8s.ginger.*` - This is the Kubernetes API URL format.

Logging in using the Giant Swarm API URL is not in the spec, and I don't believe it will bring too much value, as users don't usually interact with it directly. We can add it if necessary, nevertheless.
The error that is thrown when trying to use the Giant Swarm API URL is incorrect, and I'll be fixing that now

File: cmd/login/command.go (hunk 9 above)
Comment (corest) [reply cmd/login/command.go]: Isn’t this k8s api plugin? Why would it talk to giant swarm api?

**Comment by marians** [comment]
> Logging in using the Giant Swarm API URL is not in the spec

Right, confusion on my part.

Will re-test soon!

File: cmd/login/command.go (hunk 9 above)
Comment (marians) [reply cmd/login/command.go]: Right, it shouldn't talk to the rest API. It should also not talk to happa, but we are using the happa URL as a way for people to simply express "I want to use that control plane". But using the rest API was in fact not part of the spec. And that's fine. (I got confused about the spec)

**Comment by marians** [comment]
```
$ kgs login
Error: selected context non compatible error: The current context '' does not seem to belong to a Giant Swarm control plane.
Please run 'kgs login --help' to find out how to log in to a particular control plane.
panic: "{\"kind\":\"selectedContextNonCompatibleError\",\"annotation\":\"The current context '' does not seem to belong to a Giant Swarm control plane.\\nPlease run 'kgs login --help' to find out how to log in to a particular control plane.\",\"stack\":[{\"file\":\"/Users/marian/git/giantswarm/kubectl-gs/cmd/login/runner.go\",\"line\":94},{\"file\":\"/Users/marian/git/giantswarm/kubectl-gs/cmd/login/runner.go\",\"line\":52},{\"file\":\"/Users/marian/git/giantswarm/kubectl-gs/cmd/login/runner.go\",\"line\":40},{\"file\":\"/Users/marian/git/giantswarm/kubectl-gs/main.go\",\"line\":57}]}"
```

The current context name is missing in the error output

**Comment by axbarsan** [comment]
> The current context name is missing in the error output

Should work now 👍

**Pull request merged**